src/nbl/video/CVulkanConnection.h

```diff
@@ -42,79 +59,86 @@ class CVulkanConnection final : public IAPIConnection
         instanceCreateInfo.pNext = (VkDebugUtilsMessengerCreateInfoEXT*)&debugUtilsMessengerCreateInfo;
+        instanceCreateInfo.flags = static_cast<VkInstanceCreateFlags>(0);
         instanceCreateInfo.pApplicationInfo = &applicationInfo;
         instanceCreateInfo.enabledLayerCount = static_cast<uint32_t>(requiredInstanceLayerNames.size());
         instanceCreateInfo.ppEnabledLayerNames = requiredInstanceLayerNames.data();
+        instanceCreateInfo.enabledExtensionCount = 3u;
+        instanceCreateInfo.ppEnabledExtensionNames = instanceExtensions;
 
+        VkInstance vk_instance;
+        if (vkCreateInstance(&instanceCreateInfo, nullptr, &vk_instance) != VK_SUCCESS)
         {
-            // Todo(achal): Get this from the user
-            const bool isWorkloadHeadlessCompute = false;
+            printf("Failed to create vulkan instance, for some reason!\n");
+            return nullptr;
+        }
 
-            // Todo(achal): Not always use the debug messenger. There are also other extensions for this, check if we want to use them.
-            if (isWorkloadHeadlessCompute)
-            {
-                const uint32_t instanceExtensionCount = 1u;
-                const char* instanceExtensions[instanceExtensionCount] = { VK_EXT_DEBUG_UTILS_EXTENSION_NAME };
+        // Todo(achal): Perhaps use volkLoadInstanceOnly?
+        volkLoadInstance(vk_instance);
+        VkDebugUtilsMessengerEXT vk_debugMessenger;
+        if (vkCreateDebugUtilsMessengerEXT(vk_instance, &debugUtilsMessengerCreateInfo, nullptr, &vk_debugMessenger) != VK_SUCCESS)
+        {
+            printf("Failed to create debug messenger for some reason!\n");
+            return nullptr;
+        }
 
-                instanceCreateInfo.enabledExtensionCount = instanceExtensionCount;
-                instanceCreateInfo.ppEnabledExtensionNames = instanceExtensions;
-            }
-            else
+        constexpr uint32_t MAX_PHYSICAL_DEVICE_COUNT = 16u;
+        uint32_t physicalDeviceCount = 0u;
+        VkPhysicalDevice vk_physicalDevices[MAX_PHYSICAL_DEVICE_COUNT];
+        {
+            VkResult retval = vkEnumeratePhysicalDevices(vk_instance, &physicalDeviceCount, nullptr);
+            if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
             {
-                // Todo(achal): This needs to be handled in a platform agnostic way.
-                const uint32_t instanceExtensionCount = 3u;
-                char* instanceExtensions[instanceExtensionCount] = { "VK_KHR_surface", "VK_KHR_win32_surface", VK_EXT_DEBUG_UTILS_EXTENSION_NAME };
-
-                instanceCreateInfo.enabledExtensionCount = instanceExtensionCount;
-                instanceCreateInfo.ppEnabledExtensionNames = instanceExtensions;
+                printf("Failed to enumerate physical devices!\n");
+                return nullptr;
             }
+
+            vkEnumeratePhysicalDevices(vk_instance, &physicalDeviceCount, vk_physicalDevices);
```

Comment:
handle the case when `physicalDeviceCount>=MAX_PHYSICAL_DEVICE_COUNT` log an error or something about truncating the list

Reply:
Resolved